### ISSUE TITLE
-estwrite, alt- sometimes forgets -reg, level()-

### ISSUE THREAD

**Issue by NilsEnevoldsen** [benjann/estwrite#4]
`estwrite, alt` (or `estwrite, reproducible`) sometimes forgets the `level()` option of a regression, and instead uses the default `level(95)`. I say "sometimes" because it seems like it works if no intervening regression command is issued, though I haven't nailed down why.

Reproduction:
```stata
version 18.0
clear all
sysuse auto, clear
qui reg price weight, level(90)
estimates store est1
di _r_lb[weight] // The lower bound of the 90% CI
qui reg mpg length // If you omit this line, the issue does not reproduce
tempfile estimates
estwrite est1 using `estimates', alt // If you omit the -alt- option, the issue does not reproduce
estread `estimates'
estimates restore est1
di _r_lb[weight] // The lower bound of the 95% CI (!?)
```

Actual result:
```stata
[...]

. di _r_lb[weight]
1.4161462

[...]

. di _r_lb[weight]
1.2928575
```

Expected result:
```stata
[...]

. di _r_lb[weight]
1.4161462

[...]

. di _r_lb[weight]
1.4161462
```

I am aware of the Technical Note in the helpfile:
> It may not be possible to replay the full standard output of such a model after importing the model by estread (for example, typing regress without arguments returns error on an imported model that was fitted by regress). However, all results are, in fact, preserved and post-estimation commands such as test or estimates table should work as usual.

My understanding of the note is that `levels()` should be preserved.

The issue also reproduces with `eststo` in place of `estimates store`.

**Comment by NilsEnevoldsen** [comment]
I wonder if this is related to the following behavior of `estimates replay`. If it's not related, then apologies in advance for muddying the waters.

```stata
// -estimates replay- returns r(301); "last estimates not found"

version 18.0
clear all
sysuse auto, clear

reg price weight
tempfile foo
estwrite using `foo', alt

estread `foo'
estimates restore
estimates replay
```
```stata
// Succeeds without -estwrite …, alt-

version 18.0
clear all
sysuse auto, clear

reg price weight
tempfile foo
estwrite using `foo'

estread `foo'
estimates restore
estimates replay
```
```stata
// Succeeds using -estimates save- and -estimates use-

version 18.0
clear all
sysuse auto, clear

reg price weight
tempfile foo
estimates save `foo'

estimates use `foo'
estimates restore
estimates replay
```
```stata
// Succeeds using -estimates save- and -estimates use- even with Stata 8.0

version 8.0
clear all
sysuse auto, clear

reg price weight
tempfile foo
estimates save `foo'

estimates use `foo'
estimates restore
estimates replay
```

**Comment by benjann** [comment]
Hi Niels, `regress` does not appear to store the level in `e(level)` or similar. Also note that typing `regress` to replay the results does not take account of the original `level()` option:

```
reg price weight, level(90)
reg
```

Possibly `di _r_lb[weight]` reads results from `r(table)`, which is not stored by `estwrite, alt` (but maybe by `estimates save`, which is called by `estwrite` if the `alt` option is omitted).

ben

**Comment by NilsEnevoldsen** [comment]
The only reason I need `estwrite, alt` is because `estwrite, reproducible` implies it. I bit the bullet and decided solve all these potential problems (and possibly create others) by modifying the binary files created by `__st_estimatessave_wrk()` to replace date strings, instead of using the current `reproducible alt` implementation (https://github.com/benjann/estwrite/issues/2#issuecomment-1012592448).

In case anyone stumbling across this issue wants to try potentially shooting themselves in the foot in the same way, I invite you to consult the following solution.

```stata
tempfile estimate_temp
estwrite * using `estimate_temp'
mata {
  fi = "`estimate_temp'"
  fo = "/output/path/for/estimate.sters"
  unlink(fo)
  fh_in  = fopen(fi, "r")
  fh_out = fopen(fo, "w")
  while ((line=fgetnl(fh_in))!=J(0,0,"")) {
    if (regexm(line,"^\*! Date")==1) {
      line = regexr(line,"^\*! Date \d\d \w\w\w \d\d\d\d \d\d:\d\d:\d\d","*! Date 01 Jan 1970 00:00:00")
      line = regexr(line,"^\*! Date \d\d\w\w\w\d\d\d\d \d\d:\d\d:\d\d","*! Date 01jan1970 00:00:00")
      fwrite(fh_out, line)
    }
    else {
      fwrite(fh_out, line)
    }
  }
  fclose(fh_in)
  fclose(fh_out)
}
```

I tried this on MacOS, Stata 18.0.

Two notes:
1. It's important that the replacements be exactly the same byte length as the originals.
2. I'm not an expert at mata I/O, so I did something kind of weird. I am reading the file by line (`fgetnl()`) as though it were a text file, even though large parts of the file are binary. If a "line" begins with `*! Date` I assume that it's text and `regexr()` won't mess it up. If a "line" *doesn't* begin with `*! Date` it might be binary, and I assume that it's unsafe to pass through `regex()`. This might be a stupid method, but if it's stupid and it works…
